File: js/src/maplestory/api/dto/character/characterLinkSkillDto.ts
import { CharacterLinkSkillInfoDto } from '@src/maplestory/api/dto/character/characterLinkSkillInfoDto';
import { CharacterLinkSkillDtoBody } from '@src/maplestory/api/response/character/characterLinkSkillDtoBody';

/**
 * 캐릭터 링크 스킬 정보
 */
class CharacterLinkSkillDto {
  /**
   * 조회 기준일
   */
  date: Date;

  /**
   * 캐릭터 직업
   */
  characterClass: string;

  /**
   * 링크 스킬 정보
   */
  characterLinkSkill: CharacterLinkSkillInfoDto[];

  /**
   * 링크 스킬 1번 프리셋 정보
   */
  characterLinkSkillPreset1: CharacterLinkSkillInfoDto[];

  /**
   * 링크 스킬 2번 프리셋 정보
   */
  characterLinkSkillPreset2: CharacterLinkSkillInfoDto[];

  /**
   * 링크 스킬 3번 프리셋 정보
   */
  characterLinkSkillPreset3: CharacterLinkSkillInfoDto[];

  /**
   * 내 링크 스킬 정보
   */
  characterOwnedLinkSkill: CharacterLinkSkillInfoDto;

  /**
   * 내 링크 스킬 1번 프리셋 정보
   */
  characterOwnedLinkSkillPreset1: CharacterLinkSkillInfoDto | null;

  /**
   * 내 링크 스킬 2번 프리셋 정보
   */
  characterOwnedLinkSkillPreset2: CharacterLinkSkillInfoDto | null;

  /**
   * 내 링크 스킬 3번 프리셋 정보
   */
  characterOwnedLinkSkillPreset3: CharacterLinkSkillInfoDto | null;

  constructor(obj: CharacterLinkSkillDtoBody) {
    const {
      date,
      character_class,
      character_link_skill,
      character_link_skill_preset_1,
      character_link_skill_preset_2,
      character_link_skill_preset_3,
      character_owned_link_skill,
      character_owned_link_skill_preset_1,
      character_owned_link_skill_preset_2,
      character_owned_link_skill_preset_3,
    } = obj;

    this.date = new Date(date);
    this.characterClass = character_class;
    this.characterLinkSkill = character_link_skill.map(
      (skill) => new CharacterLinkSkillInfoDto(skill),
    );
    this.characterLinkSkillPreset1 = character_link_skill_preset_1.map(
      (skill) => new CharacterLinkSkillInfoDto(skill),
    );
    this.characterLinkSkillPreset2 = character_link_skill_preset_2.map(
      (skill) => new CharacterLinkSkillInfoDto(skill),
    );
    this.characterLinkSkillPreset3 = character_link_skill_preset_3.map(
      (skill) => new CharacterLinkSkillInfoDto(skill),
    );
    this.characterOwnedLinkSkill = new CharacterLinkSkillInfoDto(
      character_owned_link_skill,
    );
    this.characterOwnedLinkSkillPreset1 = character_owned_link_skill_preset_1 ? new CharacterLinkSkillInfoDto(
      character_owned_link_skill_preset_1,
    ) : null;
    this.characterOwnedLinkSkillPreset2 = character_owned_link_skill_preset_2 ?  new CharacterLinkSkillInfoDto(
      character_owned_link_skill_preset_2,
    ) : null;
    this.characterOwnedLinkSkillPreset3 =  character_owned_link_skill_preset_3 ? new CharacterLinkSkillInfoDto(
      character_owned_link_skill_preset_3,
    ) : null;
  }
}

export { CharacterLinkSkillDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/msea/dto/character/CharacterLinkSkillDTO.java
package dev.spiralmoon.maplestory.api.msea.dto.character;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import dev.spiralmoon.maplestory.api.Utils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.time.ZonedDateTime;
import java.util.List;

/**
 * Character equipped link skill information
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class CharacterLinkSkillDTO implements dev.spiralmoon.maplestory.api.common.dto.character.CharacterLinkSkillDTO<
        CharacterLinkSkillInfoDTO> {

    /**
     * Reference date for query (SGT, daily data with hours and minutes set to 0)
     */
    @JsonProperty("date")
    private String date;

    /**
     * Character job
     */
    @JsonProperty("character_class")
    private String characterClass;

    /**
     * Link Skill information
     */
    @JsonProperty("character_link_skill")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterLinkSkillInfoDTO> characterLinkSkill;

    /**
     * Link Skill information for preset 1
     */
    @JsonProperty("character_link_skill_preset_1")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterLinkSkillInfoDTO> characterLinkSkillPreset1;

    /**
     * Link Skill information for preset 2
     */
    @JsonProperty("character_link_skill_preset_2")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterLinkSkillInfoDTO> characterLinkSkillPreset2;

    /**
     * Link Skill information for preset 3
     */
    @JsonProperty("character_link_skill_preset_3")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterLinkSkillInfoDTO> characterLinkSkillPreset3;

    /**
     * My Link Skill information
     */
    @JsonProperty("character_owned_link_skill")
    private CharacterLinkSkillInfoDTO characterOwnedLinkSkill;

    /**
     * My Link Skill information for preset 1
     */
    @JsonProperty("character_owned_link_skill_preset_1")
    private CharacterLinkSkillInfoDTO characterOwnedLinkSkillPreset1;

    /**
     * My Link Skill information for preset 2
     */
    @JsonProperty("character_owned_link_skill_preset_2")
    private CharacterLinkSkillInfoDTO characterOwnedLinkSkillPreset2;

    /**
     * My Link Skill information for preset 3
     */
    @JsonProperty("character_owned_link_skill_preset_3")
    private CharacterLinkSkillInfoDTO characterOwnedLinkSkillPreset3;

    /**
     * Reference date for query (SGT, daily data with hours and minutes set to 0)
     */
    public ZonedDateTime getDate() {
        return date != null
                ? Utils.toZonedDateTime(this.date)
                : null;
    }
}
